File: server/service/initialize.js
const config = require('../config');
const promise = require('../module/promise');
const request = require('../module/request');
const handler = require('../module/handler');
const afterMe = require('../service/afterMe');
const mongoose = require('../module/mongoose');
const database = require('../service/database');
const model = {
    async getMongoDBProcess() {
        let command = `ps aux|grep ${config.database.bin}/mongod|grep grep -v|awk '{print $2}'`;
        return Boolean(await promise.runCommandInPromise(command));
    },
    async getMongoDBStatus() {
        let command = `${config.database.bin}/mongo --quiet --eval "db.serverStatus().ok"`;
        return await model.getMongoDBProcess() ? await promise.runCommandInPromise(command) === '1' : false;
    },
    async getOrcaFSStatus() {
        let token = await afterMe.getToken();
        let res = await request.get(config.api.orcafs.createstatus, {}, token, true);
        return !res.errorId && res.data.currentStep && res.data.currentStep === res.data.totalStep ? true : false;
    },
    async getMongoDBMasterOrNot() {
        let command = `${config.database.bin}/mongo --quiet --eval "db.isMaster().ismaster"`;
        return await model.getMongoDBProcess() ? await promise.runCommandInPromise(command) === 'true' : false;
    },
    async getOrcaFSMasterOrNot() {
        let command = `ps aux|grep orcafs-mgmtd|grep grep -v|awk '{print $2}'`;
        return Boolean(await promise.runCommandInPromise(command));
    },
    async checkClusterEnv(param) {
        let { metadataServerIPs, storageServerIPs } = param;
        let metadataServerIPsError = await Promise.all(metadataServerIPs.map(async metadata => (Object.values(await request.get(config.api.orcafs.gettoken.replace('localhost', metadata), {}, {}, true, { tokenId: '' }))[0] ? { status: '', help: '' } : { status: 'error', help: 1 })));
        let storageServerIPsError = await Promise.all(storageServerIPs.map(async storage => (Object.values(await request.get(config.api.orcafs.gettoken.replace('localhost', storage), {}, {}, true, { tokenId: '' }))[0] ? { status: '', help: '' } : { status: 'error', help: 1 })));
        let result = !Boolean(metadataServerIPsError.concat(storageServerIPsError).filter(ip => (ip.status)).length);
        return { metadataServerIPsError, storageServerIPsError, result };
    },
    async getRaidRecommendedConfiguration(param) {
        let { metadataServerIPs, storageServerIPs } = param;
        let ipList = Array.from(new Set(metadataServerIPs.concat(storageServerIPs)));
        // let diskGroup = await Promise.all(ipList.map(async ip => ({ ip, diskList: Object.assign(await afterMe.getDiskList({ ip })).data })));
        let diskGroup = ipList.map(ip => ({ ip, diskList: Array.from({ length: 28 }).map((value, index) => ({ diskName: `/dev/nvme${index}n1`, totalSpace: 11489037516 })) }));
        let metadataList = {};
        let storageList = {};
        const getConfiguration = (diskList, serviceType) => {
            let containerList = Array.from({ length: serviceType === 'metadata' ? (diskList.length > 1 ? 1 : 0) : Math.floor(diskList.length / 8) });
            let raidLevel = serviceType === 'metadata' ? 1 : 5;
            let stripeSize = 1024 * 8;
            let diskType = 'ssd';
            return containerList.map(() => {
                let disks = serviceType === 'metadata' ? diskList.splice(0, 2) : diskList.splice(0, 8);
                let totalSpace = disks.map(disk => (disk.space)).reduce((prev, next) => (prev + next));
                return { raidLevel, diskList: disks, totalSpace, stripeSize, diskType };
            });
        };
        diskGroup.forEach(item => {
            let { ip, diskList } = item;
            diskList = diskList.filter(disk => (disk.diskName.includes('nvme'))).map(disk => ({ diskName: disk.diskName, space: disk.totalSpace }));
            if (metadataServerIPs.includes(ip) && storageServerIPs.includes(ip)) {
                metadataList[ip] = diskList.length ? getConfiguration(diskList, 'metadata') : [];
                storageList[ip] = diskList.length ? getConfiguration(diskList, 'storage') : [];
            } else if (metadataServerIPs.includes(ip)) {
                metadataList[ip] = diskList.length ? getConfiguration(diskList, 'metadata') : [];
            } else if (storageServerIPs.includes(ip)) {
                storageList[ip] = diskList.length ? getConfiguration(diskList, 'storage') : [];
            }
        });
        return { metadataServerIPs: metadataList, storageServerIPs: storageList };
    },
    handleInitParam(param) {
        let { metadataServerIPs, storageServerIPs, managementServerIPs, clientIPs, enableHA, floatIPs, hbIPs, enableCustomRAID, recommendedRAID, customRAID, enableCreateBuddyGroup } = param;
        let mongodbParam = { ipList: managementServerIPs.concat(metadataServerIPs.slice(0, 1)), replicaSet: enableHA ? true : false };
        let nodeList = Array.from(new Set(managementServerIPs.concat(metadataServerIPs, storageServerIPs)));
        const handleServiceParam = (RAIDConfig, serviceType) => {
            let serviceRaidConfig = serviceType === 'metadata' ? RAIDConfig.metadataServerIPs : RAIDConfig.storageServerIPs;
            let param = [];
            if (enableCustomRAID) {
                for (let service of serviceRaidConfig) {
                    param.push({ ip: service.ip, diskGroup: service.raidList.map(raid => ({ diskList: raid.selectedDisks.map(disk => (disk.diskName)), raidLevel: raid.arrayLevel.name.toLowerCase().replace(' ', ''), stripeSize: raid.arrayStripeSize.replace(' ', '').replace('B', '').toLowerCase() })) });
                }
            } else {
                for (let service of Object.keys(serviceRaidConfig)) {
                    param.push({ ip: service, diskGroup: serviceRaidConfig[service].map(raid => ({ diskList: raid.diskList.map(disk => (disk.diskName)), raidLevel: `raid${raid.raidLevel}`, stripeSize: `${raid.stripeSize / 1024}k` })) });
                }
            }
            return param;
        };
        let metadata = { type: 'meta', hosts: handleServiceParam(enableCustomRAID ? customRAID : recommendedRAID, 'metadata') };
        let storage = { type: 'storage', hosts: handleServiceParam(enableCustomRAID ? customRAID : recommendedRAID, 'storage') };
        let mgmt = { type: 'mgmt', hosts: managementServerIPs.map((ip, index) => ({ ip, heartBeatIp: enableHA ? hbIPs[index] : '' })), floatIp: enableHA ? floatIPs[0] : '' };
        let client = { type: 'client', hosts: clientIPs.map(ip => ({ ip })) };
        let orcafsParam = [metadata, storage, mgmt, client];
        return { mongodbParam, orcafsParam, nodeList, enableCreateBuddyGroup };
    },
    async initMongoDB(param) {
        let { ipList, replicaSet } = param;
        if (!replicaSet) {
            let command = `${config.database.bin}/mongod --dbpath ${config.database.dbpath} --logpath ${config.database.logpath} --fork`;
            await promise.runCommandInPromise(command);
            await promise.runCommandInPromise(`sed -i "/MongoDB/ a ${command}" /etc/rc.local`);
        } else {
            let command = `${config.database.bin}/mongod --dbpath ${config.database.dbpath} --logpath ${config.database.logpath} --replSet ${config.database.replicaSet} --bind_ip_all --fork`;
            let conf = { _id: config.database.replicaSet, members: ipList.map((ip, index) => ({ _id: index, host: ip, priority: 2 - index })) };
            await promise.writeFileInPromise('/tmp/.initiatedb.js', `rs.initiate(${JSON.stringify(conf)})`);
            for (let ip of ipList) {
                await promise.runCommandInRemoteNodeInPromise(ip, command);
                await promise.runCommandInRemoteNodeInPromise(ip, `sed -i "/mongod/d" /etc/rc.local`);
                await promise.runCommandInRemoteNodeInPromise(ip, `sed -i "/MongoDB/ a ${command}" /etc/rc.local`);
            }
            await promise.runCommandInPromise(`${config.database.bin}/mongo /tmp/.initiatedb.js`);
            await promise.runCommandInPromise('sleep 20');
        }
    },
    async initOrcaFS(param) {
        let token = await afterMe.getToken();
        return await request.post(config.api.orcafs.createcluster, param, token, true);
    },
    async getOrcaFSInitProgress() {
        let token = await afterMe.getToken();
        return await request.get(config.api.orcafs.createstatus, {}, token, true);
    },
    async saveInitInfo(param) {
        await mongoose.connect(`mongodb://localhost/${config.database.name}`);
        for (let i of Object.keys(param)) {
            await database.addSetting({ key: config.setting[i], value: param[i] });
        }
        await database.addUser({ username: 'admin', password: handler.md5('123456') });
        await database.addLocalAuthUserGroup({ name: 'everyone', description: 'everyone' });
        await database.addSetting({ key: config.setting.snapshotSetting, value: config.snapshot });
    },
    async antiInitMongoDB(ipList) {
        let command = `killall mongod; sleep 5; rm -rf ${config.database.dbpath}/*`;
        for (let i = 0; i < ipList.length; i++) {
            if (!i) {
                await promise.runCommandInPromise(command);
                await promise.runCommandInPromise('sed -i "/mongod/d" /etc/rc.local');
            } else {
                await promise.runCommandInRemoteNodeInPromise(ipList[i], command);
                await promise.runCommandInRemoteNodeInPromise(ipList[i], 'sed -i "/mongod/d" /etc/rc.local');
            }
        }
    },
    async antiInitOrcaFS() {
        let token = await afterMe.getToken();
        return await request.post(config.api.orcafs.destroycluster, {}, token, true);
    },
    async restartServer(ipList) {
        let command = 'service orcafs-gui restart';
        for (let ip of ipList.reverse()) {
            await promise.runCommandInRemoteNodeInPromise(ip, command);
        }
    },
    async reInitMongoDB(ipList) {
        await promise.runCommandInPromise('killall mongod');
        let command = `${config.database.bin}/mongod --dbpath ${config.database.dbpath} --logpath ${config.database.logpath} --replSet ${config.database.replicaSet} --bind_ip_all --fork`;
        let conf = { _id: config.database.replicaSet, members: ipList.map((ip, index) => ({ _id: index, host: ip, priority: 2 - index })) };
        await promise.writeFileInPromise('/tmp/.initiatedb.js', `rs.initiate(${JSON.stringify(conf)})`);
        for (let ip of ipList) {
            await promise.runCommandInRemoteNodeInPromise(ip, command);
            await promise.runCommandInRemoteNodeInPromise(ip, `sed -i "/mongod/d" /etc/rc.local`);
            await promise.runCommandInRemoteNodeInPromise(ip, `sed -i "/MongoDB/ a ${command}" /etc/rc.local`);
        }
        await promise.runCommandInPromise(`${config.database.bin}/mongo /tmp/.initiatedb.js`);
        await promise.runCommandInPromise('sleep 20');
        await mongoose.connect(`mongodb://localhost/${config.database.name}`);
    }
};
module.exports = model;

File: server-unified/lib/util/initialize.py
import json
import math

from lib.module import handler, process, request
from lib.service import backend, database


def get_orcafs_status():
    response = backend.get_create_status()
    if response['errorId']:
        status = False
    else:
        data = response['data']
        status = True if data['currentStep'] and data['currentStep'] == data['totalStep'] else False
    return status


def check_cluster_env(metadatas, storages):
    def check_server(server):
        token = handler.unicode2str(request.get(
            'http://%s:9090/token/get' % server, {}, {}, {'tokenId': ''}))
        result = {'status': '', 'help': ''} if token['tokenId'] else {
            'status': 'error', 'help': 1}
        return result

    def filter_result(server):
        return True if server['status'] else False
    metadataServerIPsError = map(check_server, metadatas)
    storageServerIPsError = map(check_server, storages)
    result = not bool(len(filter(filter_result, metadataServerIPsError +
                                 storageServerIPsError)))
    return {'metadataServerIPsError': metadataServerIPsError, 'storageServerIPsError': storageServerIPsError, 'result': result}


def get_recommended_raid_configuration(metadatas, storages):
    ip_list = list(set(metadatas + storages))

    def create_disk_group(ip):
        return {'ip': ip, 'diskList': backend.get_disk_list(ip)}

    def create_mock_disk_group(ip):
        disk_list = []
        for i in range(28):
            disk_list.append(
                {'diskName': '/dev/nvme%sn1' % i, 'totalSpace': 11489037516})
        return {'ip': ip, 'diskList': disk_list}
    disk_group = map(create_mock_disk_group, ip_list)
    metadata_configuration = {}
    storage_configuration = {}

    def get_configuration(disk_list, service_type):
        disk_list_length = len(disk_list)
        container_list = []
        container_list_length = len(container_list)
        raid_level = 0
        stripe_size = 1024 * 8
        disk_type = 'ssd'
        if service_type == 'metadata':
            if disk_list_length > 1:
                container_list_length = 1
            else:
                pass
            raid_level = 1
        else:
            container_list_length = int(math.floor(disk_list_length / 8))
            raid_level = 5
        container_list = range(container_list_length)

        def map_container_list(container):
            disks = disk_list[0:2] if service_type == 'metadata' else disk_list[0:8]
            total_space = 0
            for disk in disks:
                disk_list.remove(disk)
                total_space += disk['space']
            return {'raidLevel': raid_level, 'diskList': disks, 'totalSpace': total_space, 'stripeSize': stripe_size, 'diskType': disk_type}

        return map(map_container_list, container_list)

    for item in disk_group:
        ip = item['ip']
        disk_list = item['diskList']
        disk_list = filter(lambda disk: 'nvme' in disk['diskName'], disk_list)
        disk_list = map(lambda disk: {
                        'diskName': disk['diskName'], 'space': disk['totalSpace']}, disk_list)
        disk_list_length = len(disk_list)
        if ip in metadatas and ip in storages:
            metadata_configuration[ip] = get_configuration(
                disk_list, 'metadata') if disk_list_length else []
            storage_configuration[ip] = get_configuration(
                disk_list, 'storage') if disk_list_length else []
        elif ip in metadatas:
            metadata_configuration[ip] = get_configuration(
                disk_list, 'metadata') if disk_list_length else []
        elif ip in storages:
            storage_configuration[ip] = get_configuration(
                disk_list, 'storage') if disk_list_length else []
    return {'metadataServerIPs': metadata_configuration, 'storageServerIPs': storage_configuration}


def param_handler(metadataServerIPs, storageServerIPs, managementServerIPs, clientIPs, enableHA, floatIPs, hbIPs, enableCustomRAID, recommendedRAID, customRAID, enableCreateBuddyGroup):
    def empty_list_handler(data):
        return data if data[0] else []
    clientIPs = empty_list_handler(clientIPs)
    floatIPs = empty_list_handler(floatIPs)
    hbIPs = empty_list_handler(hbIPs)

    def mongodb_param_handler(mgmts, metadatas):
        ip_list = []
        replica_set = False
        if enableHA:
            ip_list = mgmts + metadatas[0:1]
            replica_set = True
        else:
            ip_list = mgmts
        return {'ip_list': ip_list, 'replica_set': replica_set}

    def service_param_handler(raid_config, service_type):
        service_raid_config = raid_config.get('metadataServerIPs') or raid_config['metadataNodes'] if service_type == 'metadata' else raid_config.get(
            'storageServerIPs') or raid_config['storageNodes']
        param = []
        if enableCustomRAID:
            for service in service_raid_config:
                param.append({'ip': service['ip'], 'diskGroup': map(lambda raid: {'diskList': map(lambda disk: disk['diskName'], raid['selectedDisks']), 'raidLevel': raid['arrayLevel']['name'].lower(
                ).replace(' ', ''), 'stripeSize': raid['arrayStripeSize'].replace(' ', '').replace('B', '').lower()}, service['raidList'])})
        else:
            for service in dict.keys(service_raid_config):
                param.append({'ip': service, 'diskGroup': map(lambda raid: {'diskList': map(
                    lambda disk: disk['diskName'], raid['diskList']), 'raidLevel': 'raid%s' % raid['raidLevel'], 'stripeSize': '%sk' % (raid['stripeSize'] / 1024)}, service_raid_config[service])})
        return param

    def mgmt_param_handler(mgmts, heart_beat_ips):
        param = []
        if enableHA:
            param.append({'ip': mgmts[0], 'heartBeatIp': heart_beat_ips[0]})
            param.append({'ip': mgmts[1], 'heartBeatIp': heart_beat_ips[1]})
        else:
            param.append({'ip': mgmts[0], 'heartBeatIp': ''})
        return param

    metadata = {'type': 'meta', 'hosts': service_param_handler(
        customRAID if enableCustomRAID else recommendedRAID, 'metadata')}
    storage = {'type': 'storage', 'hosts': service_param_handler(
        customRAID if enableCustomRAID else recommendedRAID, 'storage')}
    mgmt = {'type': 'mgmt', 'hosts': mgmt_param_handler(
        managementServerIPs, hbIPs), 'floatIp': floatIPs[0] if enableHA else ''}
    client = {'type': 'client', 'hosts': map(lambda ip: {'ip': ip}, clientIPs)}
    mongodb_param = mongodb_param_handler(
        managementServerIPs, metadataServerIPs)
    orcafs_param = [metadata, storage, mgmt,
                    client] if len(clientIPs) else [metadata, storage, mgmt]
    node_list = {'mgmt': managementServerIPs, 'meta': metadataServerIPs,
                 'storage': storageServerIPs, 'client': clientIPs}
    param = {'metadataServerIPs': metadataServerIPs, 'storageServerIPs': storageServerIPs, 'managementServerIPs': managementServerIPs, 'clientIPs': clientIPs, 'enableHA': enableHA,
             'floatIPs': floatIPs, 'hbIPs': hbIPs, 'enableCustomRAID': enableCustomRAID, 'recommendedRAID': recommendedRAID, 'customRAID': customRAID, 'enableCreateBuddyGroup': enableCreateBuddyGroup}
    return {'mongodb_param': mongodb_param, 'orcafs_param': orcafs_param, 'node_list': node_list, 'enable_create_buddy_group': enableCreateBuddyGroup, 'param': param}


def initialize_mongodb(param):
    ip_list = param['ip_list']
    replica_set = param['replica_set']
    if not replica_set:
        command = '/usr/bin/mongod --dbpath /var/lib/mongo --logpath /var/log/mongodb/mongod.log --fork'
        process.run(command)
    else:
        command = '/usr/bin/mongod --dbpath /var/lib/mongo --logpath /var/log/mongodb/mongod.log --replSet orcafs --bind_ip_all --fork'

        def create_members_config(members):
            param = []
            for i in range(len(members)):
                param.append({'_id': i, 'host': members[i], 'priority': 2 - i})
            return param
        config = {'_id': 'orcafs', 'members': create_members_config(ip_list)}
        for ip in ip_list:
            process.run(command, ip)
        process.run('/usr/bin/mongod --quiet --eval "rs.initiate(%s)"' %
                    json.dumps(config))
    process.run('sleep 20')


def save_initialize_information(param, node_list):
    database.connect_database()
    database.create_setting('INITIALIZE-PARAMETER', param)
    database.create_setting('NODE-LIST', node_list)
    database.create_setting(
        'SNAPSHOT-SETTING', {'total': 64, 'manual': 25, 'auto': 39})
    database.create_user('admin', 'e10adc3949ba59abbe56e057f20f883e')


def deinitialize_mongodb(node_list):
    for i in range(len(node_list)):
        process.run('killall mongod', ip)
        process.run('sleep 5', ip)
        process.run('rm -rf /var/lib/mongo/*', ip)
        process.run('echo "" > /var/log/mongodb/mongod.log', ip)
